JDK-8302801: Remove fdlibm C sources

jddarcy opened the pull request:
While the review of https://github.com/openjdk/jdk/pull/12800 finishes up, I thought I'd get out for the review the next phase of the FDLIBM port: removing the FDLIBM C sources from the repo.

A repo with the changes for JDK-8302027 and this PR successful build on the default set of platform and successfully run tier 1 tests, which includes tests of the math library.

There are a few remaining references to the case-independent string "fdlibm" in the make directory and HotSpot sources. HotSpot contains a partial fork for FDLIBM (a tine of FDLIBM?) to use for intrinsics. The remaining make machinery contains logic to determine what set of gcc options can be used for the compile.

The intention of this change is to remove use of FDLIBM for the core libraries.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302801](https://bugs.openjdk.org/browse/JDK-8302801): Remove fdlibm C sources


### Reviewers
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**) ⚠️ Review applies to [417cb739](https://git.openjdk.org/jdk/pull/12821/files/417cb7393ed9068c570fa129125a668476f9d458)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12821/head:pull/12821` \
`$ git checkout pull/12821`

Update a local copy of the PR: \
`$ git checkout pull/12821` \
`$ git pull https://git.openjdk.org/jdk pull/12821/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12821`

View PR using the GUI difftool: \
`$ git pr show -t 12821`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12821.diff">https://git.openjdk.org/jdk/pull/12821.diff</a>

</details>

AlanBateman commented:
This is a great milestone to get to!  Does the comment at the top of sharedRuntimeTrig.cpp need updating?

erikj79 commented:
Could you also remove the part in `make/autoconf/flags-cflags.m4` where `FDLIBM_CFLAGS` is setup as well as the corresponding line in `make/autoconf/spec.gmk.in` where it's exported to the makefiles.

erikj79 commented on make/autoconf/buildjdk-spec.gmk.in:
```diff
@@ -82,8 +82,6 @@ JVM_LDFLAGS := @OPENJDK_BUILD_JVM_LDFLAGS@
 JVM_ASFLAGS := @OPENJDK_BUILD_JVM_ASFLAGS@
 JVM_LIBS := @OPENJDK_BUILD_JVM_LIBS@
 
-FDLIBM_CFLAGS := @OPENJDK_BUILD_FDLIBM_CFLAGS@
```
If the hotspot build still needs `FDLIBM_CFLAGS`, then this line needs to stay.

jddarcy replied:
Okay; added back.

jddarcy replied:
PS Successful Mach 5 job of default builds and tier 1 tests with this make line present.

jddarcy replied:
> PS Successful Mach 5 job of default builds and tier 1 tests with this make line present.

PPS And for extra measure as suggested by David Holmes, a tier 1 through 5 build job was also successful.

jddarcy commented:
> This is a great milestone to get to! Does the comment at the top of sharedRuntimeTrig.cpp need updating?

Updated several of the comments in the HotSpot sources and added a description of the transliteration process to StrictMath.

bplb commented on src/java.base/share/classes/java/lang/StrictMath.java:
```diff
@@ -42,8 +42,28 @@
  * algorithms, which are written in the C programming language, are
- * then to be understood as executed with all floating-point
- * operations following the rules of Java floating-point arithmetic.
+ * then to be understood to be transliterated into Java and executed
```
This change looks good (did not build and view javadoc).

jddarcy replied:
I checked that the change rendered as desired in a local docs build.

AlanBateman commented on src/hotspot/share/runtime/sharedRuntimeTrig.cpp:
```diff
@@ -28,14 +28,10 @@
+// StrictMath. Avoiding the indirect call through function
 // pointer out to libjava.so in SharedRuntime speeds these routines up
-// by roughly 15% on both Win32/x86 and Solaris/SPARC.
+// by roughly 15% on multiple architectures.
```
These functions will no longer be in libjava so maybe the "Avoiding the indirect call ..." sentence should be dropped too.

bplb reviewed (approved):
Looks fine to me.

dholmes-ora commented:
/label add hotspot

dholmes-ora reviewed (commented):
Hotspot changes are okay but I'm a bit confused about what the hotspot code will now be used for?

dholmes-ora commented:
Actually this is really my lack of understanding about the current code: why do we intrinsify `Math` but not `StrictMath`?

vnkozlov commented:
> Hotspot changes are okay but I'm a bit confused about what the hotspot code will now be used for?

`SharedRuntime::*` runtime math functions are used on platforms where there are no HW instructions or intrinsics (Zero VM). JIT compiled code may also call them in such case (or when intrinsics disabled with flag):
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/library_call.cpp#L1815

Consider them as default intrinsics for all platforms (since they are written in C). They are faster than interpreting bytecode.

They are also needed for results consistency - the same code is used by Interpreter and JITed code.

jddarcy commented:
> Actually this is really my lack of understanding about the current code: why do we intrinsify `Math` but not `StrictMath`?

In brief, the Math methods are allowed implementation flexibility in terms of their algorithm but the StrictMath methods are not. The "interesting" StrictMath methods are required to use the FDLIBM algorithms.

(One exception is StrictMath.sqrt. Since sqrt is required to be correctly rounded, there is only one correct answer for any given argument and StrictMath.sqrt can be intrinsified to a hardware sqrt instruction just like Math.sqrt can.)

jddarcy commented:
> Hotspot changes are okay but I'm a bit confused about what the hotspot code will now be used for?

I'm not 100% positive if the current __kernel_rem_pio2 code would be in use. IIRC, back when we used the fsin/fcos instructions to intrinsify sin/cos on the x87 FPU, to meet Java semantics we needed to do argument reduction into the range supported by fsin/fcos. Perhaps  __kernel_rem_pio2 is a hold-over from that time? I believe my recent sin/cos algorithms for instrincs wouldn't need to use these pathways.

AlanBateman reviewed (approved):
I don't have any other comments on this, looks good.

vnkozlov reviewed (approved):
HotSpot changes look good.

jddarcy commented:
/integrate
